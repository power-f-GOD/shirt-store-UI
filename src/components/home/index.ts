export * from './BottomBar';
export * from './Card';
