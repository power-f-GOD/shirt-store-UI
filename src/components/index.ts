export * from './Stack';
export * from './Text';
export * from './View';
export * from './Img';
export * from './Skeleton';
