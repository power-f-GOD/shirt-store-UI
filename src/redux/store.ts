import { configureStore, Dispatch, PayloadAction } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useSelector } from 'react-redux';

import { DEVELOPMENT } from 'src/constants/misc';
import { rootReducer } from './slices';

const store = configureStore({
  reducer: rootReducer,
  devTools: DEVELOPMENT,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({ serializableCheck: false })
});

export const dispatch: Dispatch<PayloadAction<any> | any> = store.dispatch;
export const getState = store.getState;

export type StateProps = ReturnType<typeof getState>;
export type StoreDispatch = ReturnType<typeof dispatch>;
export type StoreGetState = typeof getState;

// export const useTypedDispatch: () => StoreDispatch = useReduxDispatch;
export const useTypedSelector: TypedUseSelectorHook<StateProps> = useSelector;

export default store;
