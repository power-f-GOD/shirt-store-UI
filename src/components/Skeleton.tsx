import { FC, memo, useMemo } from 'react';

import { SkeletonProps } from 'src/types';
import { Stack } from '.';

const _Skeleton: FC<SkeletonProps> = ({
  size,
  width,
  height,
  type,
  borderRadius,
  count,
  variant,
  erred,
  className: _className,
  style: _style
}) => {
  const hasRoundedClass = /(^|\s)rounded/.test(_className || '');
  const hasWidthClass = /(^|\s)w-/.test(_className || '');
  const hasHeightClass = /(^|\s)h-/.test(_className || '');
  const className = `${_className || ''}${
    hasHeightClass || height ? '' : ' h-1m'
  } ${type === 'circle' ? ' rounded-full' : ''}`;
  const style = useMemo(
    () => ({
      ..._style,
      ...((height || size) && !hasHeightClass
        ? { height: height || size }
        : {}),
      ...((width || size) && !hasWidthClass ? { width: width || size } : {}),
      ...(borderRadius && !hasRoundedClass ? { borderRadius } : {})
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      // eslint-disable-next-line react-hooks/exhaustive-deps
      JSON.stringify(_style),
      borderRadius,
      hasHeightClass,
      hasRoundedClass,
      hasWidthClass,
      height,
      size,
      width
    ]
  );

  return (
    <>
      {Array(count || 1)
        .fill('')
        .map((_, i) => (
          <Stack
            key={i}
            as="span"
            className={`Skeleton relative left-0 top-0 inline-flex justify-center items-center self-center max-width-full overflow-hidden rounded-sm ${
              className || ''
            } ${variant || ''} ${erred ? 'erred' : ''}`
              .replace(/\s+/g, ' ')
              .trim()}
            style={style}
          />
        ))}
    </>
  );
};

export const Skeleton = memo(_Skeleton);
