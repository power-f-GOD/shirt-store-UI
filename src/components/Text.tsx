import { memo, useMemo, forwardRef } from 'react';

import { TextProps } from 'src/types';
import { View } from './View';

const _Text = forwardRef<HTMLElement, TextProps>(
  (
    {
      as,
      children,
      style,
      className,
      height,
      width,
      background,
      color,
      display,
      fontSize,
      overflow,
      overflowY,
      overflowX,
      alignSelf,
      fontStyle,
      fontWeight,
      overflowEllipsis,
      textAlign,
      textTransform,
      component,
      ...restProps
    },
    ref
  ): JSX.Element => {
    const Component = component || View;
    const styleCheck = JSON.stringify(style || {});

    return (
      <Component
        {...restProps}
        as={as || 'span'}
        ref={ref}
        className={`Text ${className || ''}${
          overflowEllipsis ? ' text-ellipsis' : ''
        }`.trim()}
        style={useMemo(
          () => ({
            alignSelf,
            width,
            height,
            color,
            background,
            display,
            fontSize,
            fontStyle,
            fontWeight,
            overflow,
            overflowY,
            overflowX,
            textAlign,
            textTransform,
            ...(style || {})
          }),
          // eslint-disable-next-line react-hooks/exhaustive-deps
          [
            height,
            width,
            background,
            color,
            display,
            fontSize,
            fontStyle,
            fontWeight,
            overflow,
            overflowX,
            overflowY,
            textAlign,
            textTransform,
            styleCheck
          ]
        )}>
        {children}
      </Component>
    );
  }
);

_Text.displayName = '_Text';

export const Text = memo(_Text);
