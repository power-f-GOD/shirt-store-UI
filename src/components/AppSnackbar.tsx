import React, { useCallback, memo } from 'react';
import { shallowEqual } from 'react-redux';
import {
  Snackbar as MuiSnackbar,
  Button,
  Alert,
  Slide,
  SlideProps,
  AlertTitle
} from '@mui/material';

import { dispatch, snackbar, useTypedSelector } from 'src/redux';

const _AppSnackbar = () => {
  const {
    open,
    severity,
    variant,
    message,
    label,
    duration,
    position,
    title,
    autoHide
  } = useTypedSelector((state) => state.snackbar, shallowEqual);

  const handleOnClose = useCallback(() => {
    dispatch(snackbar({ open: false }));
  }, []);

  const SlideTransition = useCallback(
    (props: SlideProps) => {
      return (
        <Slide
          {...props}
          direction={
            position === 'top' || (!position && severity !== 'info')
              ? 'down'
              : 'up'
          }
        />
      );
    },
    [position, severity]
  );

  return (
    <MuiSnackbar
      open={!!open}
      autoHideDuration={duration || 6000}
      onClose={autoHide ? handleOnClose : undefined}
      action={
        <Button color={severity} size="small" onClick={handleOnClose}>
          {label || 'OK'}
        </Button>
      }
      TransitionComponent={
        // severity === 'info' ? undefined :
        SlideTransition
      }
      anchorOrigin={
        position === 'top' || (!position && severity !== 'info')
          ? { vertical: 'top', horizontal: 'right' }
          : globalThis.innerWidth < 768
          ? { vertical: 'bottom', horizontal: 'right' }
          : undefined
      }>
      <Alert
        onClose={handleOnClose}
        severity={severity}
        variant={variant}
        className="shadow-lg">
        {title && (
          <AlertTitle className="text-capitalize mb-1">
            {typeof title === 'boolean' ? `${severity}!` : title}
          </AlertTitle>
        )}
        {message && <>{message[0].toUpperCase() + message.slice(1)}</>}
      </Alert>
    </MuiSnackbar>
  );
};

export const AppSnackbar = memo(_AppSnackbar);
