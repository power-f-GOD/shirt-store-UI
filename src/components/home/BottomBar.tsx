import { Check } from '@mui/icons-material';
import { Button } from '@mui/material';
import { memo, FC } from 'react';

import { Stack, Text } from 'src/components/shared';
import { useTypedSelector } from 'src/redux';
import { BasketProps } from 'src/types';
import { formatNumber } from 'src/utils';

const _BottomBar: FC<Pick<BasketProps, 'item_count'>> = ({ item_count }) => {
  const { actual_cost, cost, discount } =
    useTypedSelector(
      (state) => state.orders.extra,
      (a, b) => a?.cost === b?.cost
    ) || {};

  return (
    <Stack
      as="footer"
      className="sticky bottom-0 z-30 p-5 py-3 mt-20 flex-row w-full max-w-xl mx-auto justify-between items-center bg-white/90 backdrop-blur-sm border-t border-b-0 border-solid border-gray-200 rounded-tl-2xl rounded-tr-2xl anim__fadeInUpBig sm:rounded-2xl sm:bottom-8 sm:border sm:shadow-lg sm:shadow-black/5 lg:mt-24 lg:py-4"
      animationDelay={`${0.125 * 8}s`}>
      <Stack className="">
        <Text as="small" className="text-xs">
          {item_count < 0 ? 0 : item_count} item{item_count === 1 ? '' : 's'}
          {!!actual_cost && !!discount && (
            <Text as="small" className="text-red-500 text-xs ml-2">
              (-
              {formatNumber(discount * 100, {
                maximumFractionDigits: 3
              })}
              % <Text className="hidden sm:inline">discount</Text>)
            </Text>
          )}
        </Text>

        <Stack className="font-bold flex-row items-end gap-1.5">
          <Text className="text-2xl">
            {formatNumber(cost || 0, {
              currency: 'USD',
              style: 'currency'
            })}
          </Text>
          {!!actual_cost && (
            <Text className="text-base opacity-50 mb-0.5 line-through">
              {formatNumber(actual_cost, {
                currency: 'USD',
                style: 'currency'
              })}
            </Text>
          )}
        </Stack>
      </Stack>

      <Button
        variant="contained"
        className="bg-black py-3 px-5 rounded-xl hover:bg-black/70 lg:py-3.5 lg:px-6">
        Checkout <Check className="ml-1 w-5" />
      </Button>
    </Stack>
  );
};

export const BottomBar = memo(_BottomBar);
