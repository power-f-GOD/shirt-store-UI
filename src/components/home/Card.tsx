import { memo, FC, useCallback, Dispatch, useState } from 'react';
import { IconButton } from '@mui/material';
import { Add, Remove } from '@mui/icons-material';

import S from 'src/styles/home.module.scss';
import { Img, Skeleton, Stack, Text } from 'src/components/shared';
import { Action, APIShirtProps } from 'src/types';
import { IMG_BASE_URL } from 'src/constants';
import { useTypedSelector } from 'src/redux';

const _Card: FC<
  Partial<APIShirtProps> & {
    index: number;
    dispatch: Dispatch<Action>;
  }
> = (props) => {
  const { name, image_url, index, price, dispatch } = props;
  const { discount, item } = useTypedSelector(
    (state) => ({
      discount: state.orders.extra?.discount || 0,
      item: name ? state.orders.extra?.items?.[name] : null
    }),
    (a, b) => a.item?.cost === b.item?.cost && a.discount === b.discount
  );
  const [count, setCount] = useState(0);
  const { actual_cost, cost } = item || {};

  const handleChangeItem = useCallback(
    (type: Action['type']) => () => {
      setCount((prev) => {
        const isAdd = type === 'Add';

        if (!name || (!isAdd && prev! <= 0) || (isAdd && prev! >= 30)) {
          return prev;
        }

        const count = prev + (isAdd ? 1 : -1);

        // Ensure to defer setState/dispatch to prevent the (bad setState) console error
        setTimeout(() => dispatch({ type, payload: { name, count } }));

        return count;
      });
    },
    [name, dispatch]
  );

  return (
    <Stack
      as="li"
      className={`${S.Card} ${
        count ? S.added : ''
      } relative transition m-0 h-80 rounded-2xl border border-solid overflow-clip bg-black/5 border-black/5 anim__fadeInUpTiny`}
      animationDelay={`${index * 0.125}s`}>
      {image_url ? (
        <Img
          className="transition duration-300 anim__fadeIn"
          src={`${IMG_BASE_URL}${image_url}`}
          alt={`${name} shirt`}
        />
      ) : (
        <Skeleton className="w-full h-full" />
      )}
      <Stack
        className={`${S.tray} absolute bottom-0 w-full grid transition-all duration-300 grid-cols-3 flex-row p-5 items-end justify-between pt-5 bg-white/60 backdrop-blur-sm`}>
        <Text as="h2" className="col-span-2 text-xl">
          {name || <Skeleton height="1em" className="w-full" />}
        </Text>

        <Stack className="ml-auto">
          <Text as="small">Price:</Text>
          <Text className="font-bold text-3xl">
            {price ? `$${price}` : <Skeleton height="1em" className="-mb-1" />}
          </Text>
        </Stack>

        <Stack className="flex-row items-end justify-center gap-1.5 border-0 border-t border-solid pt-5 mt-5 w-full col-span-3 border-black/10">
          <Stack className="mr-auto">
            <Text as="small" className="text-xs">
              Total Amount:
            </Text>
            <Text className="font-bold text-lg truncate">
              <Text className="">${cost || 0}</Text>
              {!!actual_cost && (
                <>
                  <Text className="opacity-40 ml-2 text-base line-through">
                    ${actual_cost || 0}
                  </Text>
                  <Text as="small" className="text-red-500 text-xs ml-2">
                    (-{discount * 100}%)
                  </Text>
                </>
              )}
            </Text>
          </Stack>

          <IconButton
            className="border border-solid border-black/10 w-8 h-8"
            onClick={handleChangeItem('Remove')}>
            <Remove />
          </IconButton>
          <Text
            className="text-lg justify-center inline-flex"
            style={{ minWidth: `${count?.toString().length || 1}ch` }}>
            {count || 0}
          </Text>
          <IconButton
            className="border border-solid border-black/10 w-8 h-8"
            onClick={handleChangeItem('Add')}>
            <Add />
          </IconButton>
        </Stack>
      </Stack>
    </Stack>
  );
};

export const Card = memo(_Card);
