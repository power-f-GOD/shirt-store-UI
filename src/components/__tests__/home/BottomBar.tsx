import { render } from '@testing-library/react';
import { Provider as ReduxProvider } from 'react-redux';

import store from 'src/redux/store';
import { BottomBar } from 'src/components/home';

jest.mock('next/navigation', () => require('next-router-mock'));

it('renders BottomBar component unchanged', () => {
  const { container } = render(
    <ReduxProvider store={store}>
      <BottomBar item_count={20} />
    </ReduxProvider>
  );

  expect(container).toMatchSnapshot();
});
