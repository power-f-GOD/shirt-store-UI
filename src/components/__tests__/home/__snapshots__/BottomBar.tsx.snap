// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BottomBar unchanged 1`] = `
<div>
  <footer
    class="Stack flex sticky bottom-0 z-30 p-5 py-3 mt-20 flex-row w-full max-w-xl mx-auto justify-between items-center bg-white/90 backdrop-blur-sm border-t border-b-0 border-solid border-gray-200 rounded-tl-2xl rounded-tr-2xl anim__fadeInUpBig sm:rounded-2xl sm:bottom-8 sm:border sm:shadow-lg sm:shadow-black/5 lg:mt-24 lg:py-4"
    style="animation-delay: 1s;"
  >
    <div
      class="Stack flex flex-col "
    >
      <small
        class="Text text-xs"
      >
        20
         item
        s
      </small>
      <div
        class="Stack flex font-bold flex-row items-end gap-1.5 transition duration-300 opacity-40"
      >
        <span
          class="Text text-2xl"
        >
          $0.0
        </span>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium bg-black py-3 px-5 rounded-xl hover:bg-black/70 lg:py-3.5 lg:px-6 css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Checkout 
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium ml-1 w-5 css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </footer>
</div>
`;
